Add Guide: The Ultimate Guide to Ephemeral Dev Environment on Gitpod (#2588)
gitpod-io/website: src/routes/guides/guide-ephemeral-dev-environment-on-gitpod/index.md

Inside your workspace, do the following:

- Run `gp init`
- Open your `.gitpod.yml` file
- Add the following snippet

```yml
# List the start up tasks. Learn more https://www.gitpod.io/docs/config-start-tasks/
tasks:
  - before: |
  	function persist_gitpod_files() {
      	# Add the file paths outside of /workspace dir, that you want to presist between reboots
      	files_to_persist=(
        	"$HOME/.bash_history"
      	)
      	local _workspace_persist_dir="/workspace/.persist";
      	mkdir -p "$_workspace_persist_dir";
      	local file file_name;
      	for file in "${files_to_persist[@]}"; do {
          	mkdir -p "${file%/*}";
          	file_name="${file##*/}";
          	if test -e "$_workspace_persist_dir/$file_name"; then {
              	rm -rf "$file";
              	printf 'info: %s\n' "Overwriting $file with workspace persisted version";
              	ln -srf "$_workspace_persist_dir/${file_name}" "$file";
          	} else {
              	touch "$file";
              	cp "$file" "$_workspace_persist_dir/";
              	ln -srf "$_workspace_persist_dir/${file_name}" "$file";
          	} fi
          	unset file_name;
      	} done

>   	}

Reviewer: I am not in favor of putting this random script in `before`. It does not look in the standard ultimate guide (not set a good example).

WDYT @loujaybee?

Reviewer: For now, that script is just a temporary workaround & it's not a permanent polished solution

I don't we should talk about that script for now till the time we have separate GitHub repo with proper docs. in the readme to educate that script

scripts like this are good for community/ customer support. If they just to get it done. In public, there should be an easy & polished solution

Reviewer 2: Yeah, I'd suggest to remove this, mentioning about persisting in the `/workspace` directory with a simpler example might make more sense.

Reviewer: to persist files, we do need that kind of script. But, the way of displaying it out here (without any description). It might create some confusion for the new users

Reviewer: either way is to have a for github repo. example (with README, educating about the problem & solution) OR simply we can remove the `Persisting files between workspace restarts`  (first one is a better option)

Reviewer: Resolved in [`85464fa`](https://github.com/gitpod-io/website/pull/2588/commits/85464fad64f585cfe4f1c974b48d522c6ece5ab3)

> ---

Reviewer 2: General notes / thoughts (comment is WIP / being updated): 

1. **Tie back to ephemeral/article topic in each section** - Would be good to mention the core topic of the article "ephemeral" as we introduce each feature throughout, and how that feature ties back to the original purpose of the article. If you do a quick CMD+F for "ephemeral" across the blog you'll see what I mean. Ephemeral is mentioned a lot in the intro (awesome!) but would be good to periodically keep bringing it back to the central topic as we introduce a new topic. 

2. **Fingers on keyboard vs passive content** - Would be good to keep in mind if we expect the reader to be "fingers on keyboard" (hands on) when they're reading, or more passively scrolling on their phone. I think the content is intended to be the former, so if possible we should give: plenty of copy/paste examples, and direct links to the docs in each section so the reader can read more about topics direct from the docs.

Reviewer: > 2. Fingers on keyboard vs passive content - Would be good to keep in mind if we expect the reader to be "fingers on keyboard" (hands on) when they're reading, or more passively scrolling on their phone. I think the content is intended to be the former, so if possible we should give: plenty of copy/paste examples, and direct links to the docs in each section so the reader can read more about topics direct from the docs.

I will take a look for this one

Reviewer 2: No worries, I think it's mostly covered now, just about making sure we have the right: 

1. Links to content in docs
2. Easy copy/paste commands where possible

🙏

author: nancy-chauhan, Siddhant-K-code, axonasif
date: Friday, 19 August 2022 02:00:00 UTC
excerpt: This guide will focus on topics and concepts that can help you set up an ephemeral dev environment on Gitpod
slug: guide-ephemeral-dev-environment-on-gitpod
teaserImage: header.jpg
image: teaser.jpg

> title: Build Projects in the Gitpod Ephemeral Dev Environment — The Ultimate Guide

Reviewer 2: **Question:** Is the preferred term "dev" or "developer" environment? 

```suggestion
title: Build Projects in a Gitpod Ephemeral Dev Environment — The Ultimate Guide
```

---

<script context="module">
  export const prerender = true;
</script>

> Gitpod is a unique remote dev environment. Rather than having a single static environment, we encourage you to write your dev environment as code. So that every time you start working, you can spin up a new workspace, start coding and throw away the workspace when you're done.

Reviewer 2: ```suggestion
Gitpod is a unique remote development tool because of a specific, unique feature: ephemeral workspaces. Whilst ephemeral workspaces are a very powerful feature, it's also quite likely that it's new to you, and that's why you're here. In this guide we'll walk you step-by-step not only what ephemerality means for a development environment in the cloud, but also how Gitpod enables you to work in this totally new way of working. Welcome to the future!

First up, what is ephemerality? Rather than having a single static environment (just like your local machine) Gitpod users are encouraged to codify their dev environment as code (more on how to do this, later!). With your project codified you'll be able to spin up a new workspace, start coding and simply throw away the workspace when you're done.
```

Let me give you a few examples where ephemerality shines :

1. **Always start from a clean slate** - If your environment gets into a bad state, spin up a new one.
2. **Work on many workspaces at once** - Let's say if you're working on a feature whilst simultaneously supporting in reviewing your colleague's code, run a new workspace for reviewing and once done, it gets deleted.

> To get the most out of ephemeral workspaces, you need to configure your project in Gitpod, forget the need to set up everything again, and get started with the coding on your [favorite IDEs/ editors](https://www.gitpod.io/docs/ides-and-editors) 🚀. This ensures when you open a new workspace, all the correct tools & packages are in place, any servers or processes are running, and more. We will also dig into neat features of Gitpod, such as 'Prebuilds', which install dependencies ahead of time so that when you open a new workspace, you don't have to wait for anything to install.

Reviewer 2: **Suggestion:** Quite a long paragraph, consider breaking up. 

```suggestion
To get the most out of ephemeral workspaces, you need to configure your project in Gitpod, forget the need to set up everything again, and get started with the coding on your [favorite IDEs/ editors](https://www.gitpod.io/docs/ides-and-editors) 🚀. 

This ensures when you open a new workspace, all the correct tools & packages are in place, any servers or processes are running, and more. We will also dig into neat features of Gitpod, such as 'Prebuilds', which install dependencies ahead of time so that when you open a new workspace, you don't have to wait for anything to install.
```

---

**Suggestion:** A lot of this content feels like it's repeating what's said above, could we entirely remove these two paragraphs? What are the critical points to save if we were to keep it? 

```suggestion
```

This guide will focus on topics and concepts that can help you set up an ephemeral dev environment on Gitpod:

1. Configuring your dev environment with code
2. Faster startup by enabling prebuilds
3. Persisting env variables and files

> 

Reviewer 2: **Suggestion:** Setup a coming section with a transitional phrase ([related Tweet](https://twitter.com/loujaybee/status/1343504468835524610?s=20&t=X6J_bSM0ToEaPgiQpGe0WA)).

```suggestion
Sound good? Let's dive in and learn how to configure your development environment with Gitpod. 
```

# Configuring your dev environment with code

> To start any project, you typically need to install the dependencies and tools required to run the project. This process becomes tedious and error-prone when you repeatedly set up your dev environment when working on ephemeral workspaces. Hence it is important to automate the setup configurations and remove redundancy.

Reviewer 2: **Suggestion:** Reframe to avoid highlighting that Gitpod is tedious, but rather the friction of not using Gitpod as the product is intended (e.g. ephemeral) _could_ become tedious. 

```suggestion
To start any project, you typically need to install the dependencies and tools required to run the project. This setup process can become tedious and error-prone if you have to repeatedly set up your dev environment. If you're working on ephemeral environments, you want to be able to simply throw away an environment and start a new one, so you don't want to have any friction to simply starting a new workspace. 
```

In Gitpod, you can configure your dev environment as code with `.gitpod.yml` and `.gitpod.Dockerfile` files. It installs all the tools and dependencies you need to get started with your project. Since both these files are committed to the version control system, you can monitor changes over time. With this, your dev environment remains consistent throughout the organization.

> ## .gitpod.yml

Reviewer 2: **Suggestion:** To facilitate scanning of the post, would be good to add some context to the title here for users who are unfamiliar with the YAML file.

```suggestion
## Configuring your project with a `.gitpod.yml`
```

> 

Reviewer 2: **Suggestion:** As a visual, would be good to include an embedded example `.gitpod.yml` file

> The `.gitpod.yml` file prepares your dev environment. This file is located at the root of your project. Every time you open your project or a repository in a new workspace, all the configured tasks are executed. You can automate any manual steps you need to get started by including it in `.gitpod.yml`. You can include what dependencies and initialization commands you need to start coding.

Reviewer 2: **Suggestion**: "The `.gitpod.yml` file prepares your dev environment." and "You can automate any manual steps" are quite open statements, would be good to embed examples of what things you can put into the yml file. 

```suggestion
The `.gitpod.yml` file prepares your dev environment by automating set up like dependency installation, and specifying which application ports to open on workspace start. Every time you open your project or a repository in a new workspace, all the configured tasks are executed. The `.gitpod.yml` file lives at the root of your project. 
```

> ### How to setup first gitpod.yml ⚡️

Reviewer 2: ```suggestion
### Setting up your first `.gitpod.yml` ⚡️
```

> `youtube: fA2fpqP1xaM `

Reviewer 2: Praise: Love that we're re-using and embedding the screencasts 😍

> 1. Open your project in the new Gitpod workspace.

Reviewer 2: **Suggestion:** I think for each of these steps we should give explicit guidance of what to do, here I think context URL is the easiest / shortest thing to explain? 

```suggestion
1. Open your repo in a new Gitpod workspace. The simplest way is by appending the entire repo URL (including the protocol) to `gitpod.io/#`, e.g. `gitpod.io/#https://github.com/gitpod-io/gitpod`
```

> 2. To get started, create a `.gitpod.yml` in the root directory of your project.

Reviewer 2: **Suggestion:** I think we can omit "to get started" since it's mentioned above 🙏 
**Suggestion:** Could we give an example, copy/paste command here? 

```suggestion
2. Create a `.gitpod.yml` in the root directory of your project, e.g. `touch `.gitpod.yml`
```

>    > Tip: Instead of manually creating it, you can type `gp init` in the terminal of your Gitpod workspace. (`gp` is a CLI for Gitpod included in all Gitpod workspaces)

Reviewer 2: ```suggestion
   > Tip: Instead of manually creating the `gitpod.yml`, type `gp init` in the terminal of your Gitpod workspace. (`gp` is a CLI for Gitpod included in all Gitpod workspaces)
```

> 3. For changes to get reflected, you need to push and commit your changes.

Reviewer 2: **Suggestion:** Swap "reflected" (ambiguous) for "Gitpod detect configuration updates", and specify that it's the `gitpod.yml` that needs to be committed. 

```suggestion
3. For Gitpod to detect configuration updates, you need to push and commit your `gitpod.yml`.
```

4. Now open your project in a new workspace by prefixing your repo URL with `https://gitpod.io/#{REPO_URL}` or using the Gitpod extension.

> > Note: You can configure your project at [gitpod.io/new](https://gitpod.io/new) , with the release of [Teams & Projects](https://www.gitpod.io/docs/teams-and-projects). It lets you collaborate with your team on a project where all members with access can view prebuilds and branches for the project.

Reviewer 2: **Suggestion:** Tempted to remove this, because: 

1. The post hasn't yet introduced Teams & Projects
2. We're not articulating the value / reasoning for doing this over what is described above
3. Can be confusing to introduce different options to users without explaining the differences, often better to give one strong recommendation. 

Alternatively we could move this content lower / introduce the features here, also. 🙏 

```suggestion
```

> You can read more about `.gitpod.yml` in our [Configure section](https://www.gitpod.io/docs/config-gitpod-file) of Documentation. To learn more about `.gitpod.yml` properties check our [`.gitpod.yml` Reference documentation](https://www.gitpod.io/docs/references/gitpod-yml).

Reviewer 2: ```suggestion
You can read more about `.gitpod.yml` in the [configure](https://www.gitpod.io/docs/config-gitpod-file) documentation. For information about the `.gitpod.yml` file properties see the [`.gitpod.yml` reference documentation](https://www.gitpod.io/docs/references/gitpod-yml).
```

## Configure Dockerfile

> You can install all tools and dependencies you need for the project with Dockerfile. With Gitpod, you can use a default Docker image - `workspace-full`, or your custom Dockerfile.

Reviewer 2: **Suggestion:** Here we should answer: why Docker, as we mention "you can use Docker to configure dependencies", but we also just stated that is what the YAML file is for further up. 

[Internal discussion](https://gitpod.slack.com/archives/C020VCB0U5A/p1660895269481089)

```suggestion
In addition to supplying commands via the `.gitpod.yml`, you can also configure a Dockerfile. Using Docker can be useful if you already have a Dockerfile for your application, are familiar with the format, or want to leverage the many benefits of the Dockerfile format, such as improved caching and performance. Gitpod workspaces by default are start with an image called `workspace-full` (more on this soon), or you can configure a custom Dockerfile.
```

### Gitpod default Docker image

By default, Gitpod uses `workspace-full` for Gitpod workspaces. `workspace-full` comes pre-installed with Docker, Nix, Go, Java, Node.js, C/C++, Python, Ruby, Rust, and PHP as well as tools such as Homebrew, Tailscale, Nginx, and several more.

> If you want a specific version of any programming language or tools, choose from the other images available on dockerhub & make sure to specify the image in your `.gitpod.yml` file.

Reviewer 2: **Suggestion:** We should really link to where the image documentation is, whilst the [workspace image docs would be better in our docs](https://github.com/gitpod-io/website/issues/2187), we should [link to the workspace-images](https://github.com/gitpod-io/workspace-images) repo for the time being. 

**Thought:** We should be careful saying "any programming language" as we'll not be providing images for all languages and/or versions. I think it's better to say "some popular" languages, etc. 

```suggestion
Gitpod [provides some pre-made and optimised workspace images](https://github.com/gitpod-io/workspace-images) for various programming languages and tools. You can select an image by adding it to your `.gitpod.yml` file, for example: 
```

For example, if you want to use Java 17, specify the following in your `.gitpod.yml` file.

```yml
image: gitpod/workspace-java-17

> ```

Reviewer 2: **Praise:** This example of how to leverage the base images is nice. 

**Suggestion:** Assuming that I (as a user) want some image other than Java 17, where should I go to find what images I can have? What about learning what is included in those?

Reviewer: It's there right in #L67 & #L73

> > Note: Pin your workspace image - For Gitpod images, we recommend using a time-stamped tag for maximum reproducibility. For example, `image: gitpod/workspace-full:2022-08-17-18-37-55` (taken from the tags panel on this [dockerhub page](https://hub.docker.com/r/gitpod/workspace-full) )

Reviewer 2: ```suggestion
> Note: To avoid unexpected breaking changes, we suggest that you explicitly pin your workspace image. For Gitpod images, we recommend using a time-stamped image tag for maximum reproducibility. For example, `image: gitpod/workspace-full:2022-08-17-18-37-55` (taken from the tags panel on this [dockerhub page](https://hub.docker.com/r/gitpod/workspace-full))
```

### Bring your own Dockerfile

> If you want to use any specific service/ language, etc. You can [Bring your own (Docker) image to Gitpod](https://www.gitpod.io/blog/docker-in-gitpod).

Reviewer 2: ```suggestion
If one of the premade workspace images doesn't fit your needs, or you want to extend an image with additional configuration, such as a specific service or programming language, you can [bring your own (Docker) image to Gitpod](https://www.gitpod.io/blog/docker-in-gitpod).
```

> 1. You can configure a custom Dockerfile by adding the following in your `.gitpod.yml`

Reviewer 2: **Suggestion:** Highlight that this property is a path, and that the naming isn't a hard requirement (e.g. if they have their own Dockerfile already). 

```suggestion
1. You can configure a custom Dockerfile by adding a reference to the Dockerfile to use in your `.gitpod.yml`. We commend the `.gitpod.Dockerfile` naming convention. 
```

```yml
image:
  file: .gitpod.Dockerfile
```

2. Next, create a `.gitpod.Dockerfile`, file at the root of your project.

> 3. You can start by keeping a base image as [gitpod/workspace-base](https://github.com/gitpod-io/workspace-images/blob/481f7600b725e0ab507fbf8377641a562a475625/dazzle.yaml#L3) image. This is an image without the default tooling, else you can go with `gitpod/workspace-full` base image.

Reviewer 2: **Thought:** This links to a Dazzle configuration, is this the best place? Maybe the README is more user friendly? 

```suggestion
3. You can start by using the lightweight base image [gitpod/workspace-base](https://github.com/gitpod-io/workspace-images) for your custom Docker image. The `workspace-base` image has some default tooling, whereas `gitpod/workspace-full` already has a lot of preinstalled installed tools.
```

### Bring your own Dockerfile

If one of the premade workspace images doesn't fit your needs, or you want to extend an image with additional configuration, such as a specific service or programming language, you can [bring your own (Docker) image to Gitpod](https://www.gitpod.io/blog/docker-in-gitpod).

1. You can configure a custom Dockerfile by adding a reference to the Dockerfile location to use in your `.gitpod.yml`. We recommend the `.gitpod.Dockerfile` naming convention. 

```yml
image:
  file: .gitpod.Dockerfile
```

2. Next, create a `.gitpod.Dockerfile`, file at the root of your project.
3. You can start by keeping a base image as [gitpod/workspace-base](https://github.com/gitpod-io/workspace-images/blob/481f7600b725e0ab507fbf8377641a562a475625/dazzle.yaml#L3) image. This is an image without the default tooling, else you can go with `gitpod/workspace-full` base image.
4. Start adding the commands you need for your dev environment.

You can read more about it in our [configure a custom Dockerfile](https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile) documentation.

# Faster startup by enabling prebuilds

> Every time you create a new ephemeral workspace, a lot of time can be saved if you use a ‘prebuild image’ of tools and dependencies. Gitpod has the feature ‘Prebuilds’. By enabling it, you will not have to wait for everything. You can start working as all the dependencies will get installed before starting a new workspace for the project. When collaborating with peers and working on the same branch, Prebuild is quite helpful as Gitpod triggers it whenever there is a commit to the repository. Hence Prebuild will be used for all the new workspaces opened on that branch.

Reviewer 2: **Suggestion:** I think we need a lead in paragraph talking about the "current" way that a developer might work, e..g on your local machine you have a single mutable environment, whereas Gitpod you can have many. I think without this context the idea of prebuilds doesn't make sense. 

**Suggestion:** "By enabling it, you will not have to wait for everything." is quite ambiguous. To explain prebuilds it can sometimes help to draw the analogous example of Continuous Integration, which might help readers understand the concept via a familiar / related concept. A key word for prebuilds is also "asynchronous" to highlight that the building happens separate to workspace starts, and is triggered separately. 

```suggestion
When working on your local machine, you typically have a single environment to configure and work on at a time, manually installing dependencies and system configurations when you switch between branches or projects. With Gitpod you can have many concurrent workspaces, and rather than doing tasks like dependency installation when you swap branch, you can start a new workspace based on your latest development environment image. 

Every time you create a new ephemeral workspace, a lot of time can be saved by using a Prebuild. Prebuilds allow you to configure asynchronous builds of your workspace image when it's updated or changed, this is a very similar approach to how Continuous Integration tools work. When collaborating with peers and working on the same branch, Prebuild is quite helpful as Gitpod triggers it whenever there is a commit to the repository. 
```

> Here is a example of project where huge time is saved using Gitpod prebuilds

Reviewer 2: **Suggestion:** For accessibility reasons, I'd suggest to write up what the content shows in the tweet, and possibly give a little added context. 

```suggestion
For example, the `.gitpod.yml` saves 41 minutes of installation and building time for the [Node.js](https://github.com/nodejs/node) repo that would otherwise have to be ran on your own machine. In Gitpod the workspace image is prebuilt, so all you have to do is start a new workspace. 
```

<figure class="flex flex-col items-center text-center">

>   <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/tweet-image.png" alt="This is tweet about of Node.js project. With Gitpod Prebuilds it got even faster" width="400" />

Reviewer 2: ```suggestion
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/tweet-image.png" alt="With Gitpod Prebuilds the Node.js project saves 41 mins of installation and build time" width="400" />
```

   <figcaption class="w-full">This is <a href="https://twitter.com/Siddhant_K_code/status/1548143768301318144">tweet</a> about Node.js project. With <a href="https://www.gitpod.io/docs/prebuilds">Gitpod Prebuilds</a> it got even faster.</figcaption>
</figure>

<figure class="flex flex-col items-center text-center">
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/gitpod-prebuilds.png" alt="Average of Time we save through prebuilds with our own Gitpod project"/>
   <figcaption class="w-full">Average of Time we save through prebuilds with our own Gitpod project.</figcaption>
</figure>

### How to get started with Prebuilds

> 1. To enable Gitpod Prebuilds for a repository, you must install the [Gitpod GitHub App](https://github.com/apps/gitpod-io) for your repository.

Reviewer 2: **Suggestion:** Note the reasoning for the GitHub app. 

```suggestion
1. To enable Gitpod Prebuilds for a repository, you must install the [Gitpod GitHub App](https://github.com/apps/gitpod-io) for your repository. The GitHub app is needed to notify Gitpod when changes are made to the repository. 
```

2. Click 'Configure' and choose the repository in which you wish to install the Gitpod App.

> 3. It creates the prebuild for every merged commit (on the particular branch), or you can configure it for every PR, which saves the prebuilt image. With this, whenever a new workspace is created, it loads the previously built image, which saves resources & time.

Reviewer 2: **Suggestion:** Swap `it` and `this` for the object in question. 

```suggestion
3. The GitHub app creates the Prebuild for every merged commit (on the particular branch). Alternatively, you can configure Preuilds for every PR. With Prebuilds, whenever a new workspace is created, Gitpod loads the previously built image, saving resources & time.
```

> To enable Prebuilds for Gitlab and Bitbucket repository check our [Prebuilds documentation](https://www.gitpod.io/docs/prebuilds#configuring-prebuilds-manually).

Reviewer 2: **Suggestion:** I'd move this qualifier above the bullet list above otherwise it can look like it's GitHub only, and mention that the example is specific to GitHub and signpost the link for installing in GitLab, etc.

> To enable Prebuilds for Gitlab and Bitbucket repository check our [Prebuilds documentation](https://www.gitpod.io/docs/prebuilds#configuring-prebuilds-manually).

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion

To enable Prebuilds for Gitlab and Bitbucket repository check our [Prebuilds documentation](https://www.gitpod.io/docs/prebuilds#configuring-prebuilds-manually).

Here is how you can enable Prebuilds for GitHub:

1. To enable Gitpod Prebuilds for a repository, you must install the [Gitpod GitHub App](https://github.com/apps/gitpod-io) for your repository.
2. Click 'Configure' and choose the repository in which you wish to install the Gitpod App.
3. It creates the prebuild for every merged commit (on the particular branch), or you can configure it for every PR, which saves the prebuilt image. With this, whenever a new workspace is created, it loads the previously built image, which saves resources & time.

```

> Here is a example where [Dropwizard](https://github.com/dropwizard/dropwizard/blob/HEAD/.gitpod.yml) has enabled the Prebuilds.The github prebuilds section in the .gitpod.yml file configures when prebuilds are run

Reviewer 2: ```suggestion
By default, Gitpod prepares prebuilt workspaces for all changes on the default branch and for pull/merge requests coming from the same repository. Here is a example where [Dropwizard](https://github.com/dropwizard/dropwizard/blob/HEAD/.gitpod.yml) has enabled the Prebuilds.The github prebuilds section in the .gitpod.yml file configures the specific conditions when prebuilds should be ran
```

```yml
github:
  prebuilds:
    # enable for the default branch
    master: true
    # enable for all branches in this repo
    branches: true
    # enable for pull requests coming from this repo
    pullRequests: true
    # enable for pull requests coming from forks
    pullRequestsFromForks: true
    # add a check to pull requests
    addCheck: true
    # add a "Review in Gitpod" button as a comment to pull requests
    addComment: false
    # add a "Review in Gitpod" button to the pull request's description
    addBadge: false
```

> > Note: Prebuilds only save the workspace directory. Any file changes made outside of `/workspace` file hierarchy from `init` tasks will be lost on workspace start when prebuilds are enabled.

Reviewer 2: **Suggestion:** This often catches folks out, would suggest to upgrade to "important" from "note".

```suggestion
> **Important:** Prebuilds only save the workspace directory. Any file changes made outside of `/workspace` file hierarchy from `init` tasks will be lost on workspace start when prebuilds are enabled.
```

slug: guide-ephemeral-dev-environment-on-gitpod
teaserImage: header.jpg
image: teaser.jpg
title: Build Projects in a Gitpod Ephemeral Dev Environment — The Ultimate Guide
---

<script context="module">
  export const prerender = true;
</script>

Gitpod is a unique remote development tool because of a specific, unique feature: ephemeral workspaces. Whilst ephemeral workspaces are a very powerful feature, it's also quite likely that it's new to you, and that's why you're here. In this guide we'll walk you step-by-step not only what ephemerality means for a development environment in the cloud, but also how Gitpod enables you to work in this totally new way of working. Welcome to the future!

First up, what is ephemerality? Rather than having a single static environment (just like your local machine) Gitpod users are encouraged to codify their dev environment as code (more on how to do this, later!). With your project codified you'll be able to spin up a new workspace, start coding and simply throw away the workspace when you're done.

Let me give you a few examples where ephemerality shines :

> 1. **Always start from a clean slate** - If your environment gets into a bad state, spin up a new one.

Reviewer 2: ```suggestion
1. **Always start from a clean slate** - If you break something in your environment rather than waste time trying to figure out exactly which environment variable or configuration you accidentally updated, simply throw away the workspace and start a brand new working workspace. 
```

> 2. **Work on many workspaces at once** - Let's say if you're working on a feature whilst simultaneously supporting in reviewing your colleague's code, run a new workspace for reviewing and once done, it gets deleted.

Reviewer 2: ```suggestion
2. **Work on many workspaces at once** - Lets say that you're working on a feature whilst simultaneously supporting in reviewing a colleague's code, rather than have to swap your local environment you can create a new parallel workspace and code in two (or more!) workspaces at once. 
```

> To get the most out of ephemeral workspaces, you need to configure your project in Gitpod, forget the need to set up everything again, and get started with the coding on your [favorite IDEs/ editors](https://www.gitpod.io/docs/ides-and-editors) 🚀.

Reviewer 2: **Suggestion:** Could we remove? Not sure what this statement adds. 

```suggestion
```

> This ensures when you open a new workspace, all the correct tools & packages are in place, any servers or processes are running, and more. We will also dig into neat features of Gitpod, such as 'Prebuilds', which install dependencies ahead of time so that when you open a new workspace, you don't have to wait for anything to install.

Reviewer 2: ```suggestion
Gitpod ensures when you open a new workspace, all the correct tools & packages are in place, any servers or processes are running, and more. We will also dig into neat features of Gitpod, such as 'Prebuilds', which install dependencies ahead of time so that when you open a new workspace, you don't have to wait for anything to install.
```

This guide will focus on topics and concepts that can help you set up an ephemeral dev environment on Gitpod:

1. Configuring your dev environment with code
2. Faster startup by enabling prebuilds
3. Persisting env variables and files

Sound good? Let's dive in and learn how to configure your development environment with Gitpod.

# Configuring your dev environment with code

To start any project, you typically need to install the dependencies and tools required to run the project. This setup process can become tedious and error-prone if you have to repeatedly set up your dev environment. If you're working on ephemeral environments, you want to be able to simply throw away an environment and start a new one, so you don't want to have a single barrier to simply starting a new workspace.

> In Gitpod, you can configure your dev environment as code with `.gitpod.yml` and `.gitpod.Dockerfile` files. It installs all the tools and dependencies you need to get started with your project. Since both these files are committed to the version control system, you can monitor changes over time. With this, your dev environment remains consistent throughout the organization.

Reviewer 2: ```suggestion
In Gitpod, you can configure your dev environment as code with `.gitpod.yml` and `.gitpod.Dockerfile` files. Gitpod installs all the tools and dependencies you need to get started with your project. Since both these files are committed to the version control system, you can monitor changes over time. With this, your dev environment remains consistent throughout the organization.
```

> In Gitpod, you can configure your dev environment as code with `.gitpod.yml` and an optional `.gitpod.Dockerfile` file. Gitpod can be configured to install all the tools and dependencies you need to get started with your project. Since both these configuration files are committed to the version control system, you get all of the benefits of source control, such as viewing the history and authors of files. With this configuration, your dev environment remains consistent throughout the organization.

Reviewer 2: ```suggestion
In Gitpod, you can configure your dev environment as code with `.gitpod.yml` and an optional `.gitpod.Dockerfile` file. Gitpod can be configured to install all the tools and dependencies you need to get started with your project. Since both these configuration files are committed to version control you get all of the benefits of source control, such as viewing the history and authors of files. With this configuration, your dev environment remains consistent throughout the organization.
```

## Configuring your project with a `.gitpod.yml`

The `.gitpod.yml` file prepares your dev environment by automating set up like dependency installation, and specifying which application ports to open on workspace start. Every time you open your project or a repository in a new workspace, all the configured tasks are executed. The `.gitpod.yml` file lives at the root of your project. For example:

```yml
# Commands to start on workspace startup
tasks:
  - init: yarn install
    command: yarn build
# Ports to expose on workspace startup
ports:
  - port: 8000
    onOpen: open-preview
```

You can read more about `.gitpod.yml` in [docs: config gitpod file](https://www.gitpod.io/docs/config-gitpod-file)

### Setting up your first `.gitpod.yml` ⚡️

`youtube: fA2fpqP1xaM `

1. Open your repo in a new Gitpod workspace. The simplest way is by appending the entire repo URL (including the protocol) to `gitpod.io/#`, e.g. `gitpod.io/#https://github.com/gitpod-io/gitpod`

> 2. Create a `.gitpod.yml` in the root directory of your project, e.g. `touch `.gitpod.yml`

Reviewer 2: ```suggestion
2. Create a `.gitpod.yml` in the root directory of your project, e.g. `touch .gitpod.yml`
```

   > Tip: Instead of manually creating the `gitpod.yml`, type `gp init` in the terminal of your Gitpod workspace. (`gp` is a CLI for Gitpod included in all Gitpod workspaces)
3. For Gitpod to detect configuration changes, push and commit your `gitpod.yml`.
4. Now open your project in a new workspace by prefixing your repo URL with `https://gitpod.io/#{REPO_URL}` or using the Gitpod extension.

You can read more about `.gitpod.yml` in the [configure](https://www.gitpod.io/docs/config-gitpod-file) documentation. For information about the `.gitpod.yml` file properties see the [`.gitpod.yml` reference documentation](https://www.gitpod.io/docs/references/gitpod-yml).

> ## Configure Dockerfile

Reviewer 2: ```suggestion
## Configure your developer environment based on a Dockerfile
```

### Bring your own Dockerfile

If one of the premade workspace images doesn't fit your needs, or you want to extend an image with additional configuration, such as a specific service or programming language, you can [bring your own (Docker) image to Gitpod](https://www.gitpod.io/blog/docker-in-gitpod).

1. You can configure a custom Dockerfile by adding a reference to the Dockerfile location to use in your `.gitpod.yml`. We recommend the `.gitpod.Dockerfile` naming convention.

```yml
image:
  file: .gitpod.Dockerfile
```

2. Next, create a `.gitpod.Dockerfile`, file at the root of your project.
3. You can start by using the lightweight base image [gitpod/workspace-base](https://github.com/gitpod-io/workspace-images) for your custom Docker image. The `workspace-base` image has some default tooling, whereas `gitpod/workspace-full` already has a lot of preinstalled installed tools.
4. Start adding the commands you need for your dev environment.

> You can read more about it in our [configure a custom Dockerfile](https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile) documentation.

Reviewer 2: ```suggestion
Read more about how to [configure a custom Dockerfile](https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile).
```

For example, the `.gitpod.yml` saves 41 minutes of installation and building time for the [Node.js](https://github.com/nodejs/node) repo that would otherwise have to be ran on your own machine. In Gitpod the workspace image is prebuilt, so all you have to do is start a new workspace. 

<figure class="flex flex-col items-center text-center">
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/tweet-image.png" alt="This is tweet about of Node.js project. With Gitpod Prebuilds it got even faster" width="400" />
   <figcaption class="w-full">This is <a href="https://twitter.com/Siddhant_K_code/status/1548143768301318144">tweet</a> about Node.js project. With <a href="https://www.gitpod.io/docs/prebuilds">Gitpod Prebuilds</a> it got even faster.</figcaption>
</figure>

<figure class="flex flex-col items-center text-center">
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/gitpod-prebuilds.png" alt="Average of Time we save through prebuilds with our own Gitpod project"/>
   <figcaption class="w-full">Average of Time we save through prebuilds with our own Gitpod project.</figcaption>
</figure>

### How to get started with Prebuilds

> To enable Prebuilds for Gitlab and Bitbucket repository check our [Prebuilds documentation](https://www.gitpod.io/docs/prebuilds#configuring-prebuilds-manually).
> 
> Here is how you can enable Prebuilds for GitHub:

Reviewer 2: ```suggestion
Prebuilds work for GitHub, Gitlab and Bitbucket repositories (see our [Prebuilds documentation](https://www.gitpod.io/docs/prebuilds#configuring-prebuilds-manually) for more). To give you an example, let's take a look at how you can enable Prebuilds for GitHub:

```

1. To enable Gitpod Prebuilds for a repository, you must install the [Gitpod GitHub App](https://github.com/apps/gitpod-io) for your repository.
2. Click 'Configure' and choose the repository in which you wish to install the Gitpod App.

> 3. It creates the prebuild for every merged commit (on the particular branch), or you can configure it for every PR, which saves the prebuilt image. With this, whenever a new workspace is created, it loads the previously built image, which saves resources & time.

Reviewer 2: ```suggestion
3. The GitHub app will create the prebuild for every merged commit (on the particular branch), or you can configure it for every PR, which saves the prebuilt image. With this, whenever a new workspace is created, it loads the previously built image, which saves resources & time.
```

    # enable for pull requests coming from forks
    pullRequestsFromForks: true
    # add a check to pull requests
    addCheck: true
    # add a "Review in Gitpod" button as a comment to pull requests
    addComment: false
    # add a "Review in Gitpod" button to the pull request's description
    addBadge: false
```

> **Important:** Prebuilds only save the workspace directory. Any file changes made outside of `/workspace` file hierarchy from `init` tasks will be lost on workspace start when prebuilds are enabled.

You can read more about prebuilds in the [following doc](https://www.gitpod.io/docs/prebuilds).

# Persisting environment variables & files

Every time you start a project, you require API keys, secrets like database passwords or sensitive information that can't be committed to the version control system. These are usually added as environment variables that are used to configure a value in your code. When working with an ephemeral workspace, you will have to set up everything again for each new task. We will discuss a few ways with which we can persist these environment variables across Gitpod workspaces.

## Configure environment variables in Gitpod

> You can configure the persistent environment variables in your Gitpod [account settings](https://gitpod.io/variables).

Reviewer 2: ```suggestion
You can configure persistent environment variables in your Gitpod [account settings](https://gitpod.io/variables).
```

<figure class="flex flex-col items-center text-center">
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/env-vars.png" alt="Configure the persistent environment variables in your Gitpod account settings." />
</figure>

The scope of each variable determines in what workspaces it will be available. Scope patterns follow the `owner/repository` pattern:

> - You can use a wildcard on either of the two, e.g. `dropwizard/*` would make that variable available in all repositories owned by dropwizard.

Reviewer 2: **Suggestion:** Add a note on the `*` pattern for users who might not be familiar with the pattern. 

```suggestion
- You can use a wildcard (denoted with a `*`) to match all strings to either of the `owner` or `repo`, e.g. `dropwizard/*` would make that environment variable available in all repositories owned by dropwizard.
```

> - You can use a wildcard on either of the two, e.g. `dropwizard/*` would make that variable available in all repositories owned by dropwizard.
> - Conversely `*/dropwizard` would make that variable available on all repositories called `dropwizard`; this is especially useful for forks.
> - Subsequently `*/*` makes that variable available in every workspace.

Reviewer 2: ```suggestion
You can use a wildcard (denoted with a `*`) to match of either the `owner` or `repo`, for example: 

- **Single organisations** - Setting `dropwizard/*` would make an environment variable available in all repositories owned by the organisation dropwizard.
- **Single repository** - `*/dropwizard` would make an environment variable available on all repositories called `dropwizard`; this is especially useful for forks.
- **All organisations or repositories** - Using `*/*` would make an environment variable available in every repository for any organisation for that given user.
```

<figure class="flex flex-col items-center text-center">
  <img src="/images/guides/guide-ephemeral-dev-environment-on-gitpod/scope.png" alt="Scope of each variable determines in what workspaces it will be available" width="500" />
</figure>

> You can read more about environment variables in the [Environment Variables on Gitpod](https://www.gitpod.io/docs/environment-variables).

Reviewer 2: ```suggestion
You can read more about environment variables in [Environment Variables on Gitpod](https://www.gitpod.io/docs/environment-variables).
```

> ### <i>(Optional)</i> Persisting files between workspace reboots

Reviewer 2: **Suggestion:** Some of the other content is "optional", do we need a specific title here for that? 

```suggestion
### Persisting files between workspace reboots
```

Reviewer: following solution under this title is not reliable (& not an optimal one) & I don't think, all folks would need to persist some X files. Rest all the required steps, I think!

Reviewer 2: Not sure I agree. Environment variables, Docker, even Prebuilds could be considered optional? If we're not confident of the content, I'd suggest we update the content, not caveat it. If we are going to caveat, I'd be explicit about the caveat, i.e mention the shortcomings of the solution, rather than mark the section as optional.

Reviewer: It was marked as Optional because of this https://github.com/gitpod-io/website/pull/2588#discussion_r948750283

I will remove that script part. So, it would be good to go. (without marking it as optional)

> There would be some use cases where we would want to persist some files through the reboot. For example, `.bash_history`, to later audit user activity.

Reviewer 2: **Suggestion:** Swap reboot terminology for restart (reboot implies a machine / VM)

```suggestion
There would be some use cases where we would want to persist some files between workspace starts. For example, `.bash_history`, to later audit user activity.
```

- Run `gp init`
- Open your `.gitpod.yml` file
- Add the following snippet

```yml
# List the start up tasks. Learn more https://www.gitpod.io/docs/config-start-tasks/
tasks:
  - before: |
  	function persist_gitpod_files() {
      	# Add the file paths outside of /workspace dir, that you want to presist between reboots
      	files_to_persist=(
        	"$HOME/.bash_history"
      	)
      	local _workspace_persist_dir="/workspace/.persist";
      	mkdir -p "$_workspace_persist_dir";
      	local file file_name;
      	for file in "${files_to_persist[@]}"; do {
          	mkdir -p "${file%/*}";
          	file_name="${file##*/}";
          	if test -e "$_workspace_persist_dir/$file_name"; then {
              	rm -rf "$file";
              	printf 'info: %s\n' "Overwriting $file with workspace persisted version";
              	ln -srf "$_workspace_persist_dir/${file_name}" "$file";
          	} else {
              	touch "$file";
              	cp "$file" "$_workspace_persist_dir/";
              	ln -srf "$_workspace_persist_dir/${file_name}" "$file";
          	} fi
          	unset file_name;
      	} done
  	}
```

# Recommended Reading

> - [Docs: Config Dotfiles](https://www.gitpod.io/docs/config-dotfiles)
> 
> * [GitHub Repository: Bring your own Dotfiles to Gitpod | dotfiles on Gitpod](https://github.com/gitpod-io/demo-dotfiles-with-gitpod)

Reviewer 2: **Suggestion:** Mixed formatting and space doesn't seem to cause an issue, but might as well join up (if that's the intention). 

```suggestion
* [Docs: Config Dotfiles](https://www.gitpod.io/docs/config-dotfiles)
* [GitHub Repository: Bring your own Dotfiles to Gitpod | dotfiles on Gitpod](https://github.com/gitpod-io/demo-dotfiles-with-gitpod)
```

* [GitHub Repository: Bring your own Dotfiles to Gitpod | dotfiles on Gitpod](https://github.com/gitpod-io/demo-dotfiles-with-gitpod)
* [GitHub Repository: AWS Single Sign-On (SSO) and Amazon Elastic Container Registry (ECR) with Gitpod](https://github.com/gitpod-io/demo-aws-with-gitpod)
* [Blog: Gitpodifying — The Ultimate Guide](https://www.gitpod.io/blog/gitpodify)
* [Blog: Personalize your Gitpod Workspace Environments](https://www.gitpod.io/blog/personalize-your-gitpod-workspace-environment)

---

# Ready to code 🚀

> Yay! You are now ready to get started with the ephemeral Dev Environment on Gitpod.

Reviewer 2: ```suggestion
Yay! You are now ready to get started with an ephemeral Dev Environment on Gitpod.

```
